forecasting/src/autogluon/forecasting/trainer/abstract_trainer.py

```diff
@@ -0,0 +1,668 @@
+        explicit_order = ["model", "val_score", "fit_order"]
+        explicit_order = [
+            column for column in explicit_order if column in df_columns_lst
+        ]
+        df_columns_other = [
+            column for column in df_columns_lst if column not in explicit_order
+        ]
+        df_columns_new = explicit_order + df_columns_other
+        df_sorted = df_sorted[df_columns_new]
+
+        return df_sorted
+
+    def _get_model_for_prediction(
+        self, model: Optional[Union[str, AbstractForecastingModel]] = None
+    ) -> AbstractForecastingModel:
+        """Given an optional identifier or model object, return the model to perform predictions
+        with. If the model is not provided, this method will default to the best model according to
+        the validation score."""
+
+        if model is None:
+            logger.log(
+                30,
+                "Model not specified, will default to the model with the best validation score",
+            )
+            if self.model_best is None:
+                best_model_name: str = self.get_model_best()
+                self.model_best = best_model_name
+            return self.load_model(self.model_best)
+        else:
+            if isinstance(model, str):
+                return self.load_model(model)
+            return model
+
+    def predict(
+        self,
+        data: TimeSeriesDataFrame,
+        model: Optional[AbstractForecastingModel] = None,
+        **kwargs,
+    ) -> TimeSeriesDataFrame:
+        model = self._get_model_for_prediction(model)
+        return self._predict_model(data, model, **kwargs)
+
+    def score(
+        self,
+        data: TimeSeriesDataFrame,
+        model: Optional[Union[str, AbstractModel]] = None,
+        quantiles: List[float] = None,
+    ) -> float:
+        evaluator = Evaluator(quantiles=quantiles or self.quantiles)
+        model = self._get_model_for_prediction(model)
+
+        # FIXME: this method should be able to score on all data sets regardless of
+        # FIXME: whether the implementation is in GluonTS
+        if not isinstance(model, AbstractGluonTSModel):
+            raise ValueError("Model must be a GluonTS model to predict for scoring")
+        forecasts, tss = model._predict_for_scoring(data)  # noqa
+        num_series = len(tss)
+
+        with evaluator_warning_filter():
+            agg_metrics, item_metrics = evaluator(
+                iter(tss), iter(forecasts), num_series=num_series
+            )
+        return agg_metrics[self.eval_metric]
+
+    def _predict_model(
+        self,
+        data: TimeSeriesDataFrame,
+        model: Union[str, AbstractForecastingModel],
+        **kwargs,
+    ) -> TimeSeriesDataFrame:
+        if isinstance(model, str):
+            model = self.load_model(model)
+        return model.predict(data, **kwargs)
+
+    # TODO: experimental
+    def refit_single_full(self, train_data=None, val_data=None, models=None):
```

Comment:
Do we think it is possible to make this in SimpleAbstractTrainer or some kind of mixin? The logic for refit full is pretty generic overall, and I foresee that we could maybe clean up the code to make it do-able as fully generic.

Reply:
The problem is that `train_data` and `val_data` conflict with the `X, y` based signatures of core. So whenever there is a function with data arguments we have to override it in forecast. I think this is part of a bigger issue which would be great to tackle to make this and other logic more generic.

Reply:
Ok, sounds good, definitely a bigger change so we can think about it in a future PR.